pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java

```diff
@@ -0,0 +1,323 @@
+            Path checkpointDirPath = checkpointRootPath.resolve(checkpointName);
+            Path checkpointPath = checkpointDirPath.resolve(PRAVEGA_CHECKPOINT_FILE_NAME);
+            log.info("Reading Pravega checkpoint from {}", checkpointPath);
+            Checkpoint checkpoint = Checkpoint.fromBytes(ByteBuffer.wrap(Files.readAllBytes(checkpointPath)));
+            log.info("Starting from checkpointName={}, positions={}", checkpointName, checkpoint.asImpl().getPositions());
+            builder = builder.startFromCheckpoint(checkpoint);
+            startFromCheckpointName = checkpointName;
+        } else {
+            log.warn("Checkpoint file {} not found. Starting processing from the earliest event.", checkpointRootPath);
+
+            // Create streams.
+            try (StreamManager streamManager = StreamManager.create(controllerURI)) {
+                streamManager.createScope(scope);
+                StreamConfiguration streamConfig = StreamConfiguration.builder()
+                        .scalingPolicy(ScalingPolicy.byEventRate(
+                                Parameters.getTargetRateEventsPerSec(),
+                                Parameters.getScaleFactor(),
+                                Parameters.getMinNumSegments()))
+                        .build();
+                streamManager.createStream(scope, inputStreamName, streamConfig);
+                streamManager.createStream(scope, outputStreamName, streamConfig);
+            }
+
+            // Create a reader group that starts from the earliest event.
+            builder = builder.stream(Stream.of(scope, inputStreamName));
+            startFromCheckpointName = null;
+        }
+
+        clientFactory = ClientFactory.withScope(scope, controllerURI);
+        writer = clientFactory.createEventWriter(
+                outputStreamName,
+                new UTF8StringSerializer(),
+                EventWriterConfig.builder().build());
+
+        // Create a reader group manager.
+        // It must remain open so that readerGroup.initiateCheckpoint returns.
+        readerGroupManager = ReaderGroupManager.withScope(scope, controllerURI);
+
+        final ReaderGroupConfig readerGroupConfig = builder.build();
+        readerGroupName = UUID.randomUUID().toString().replace("-", "");
+        readerGroupManager.createReaderGroup(readerGroupName, readerGroupConfig);
+        readerGroup = readerGroupManager.getReaderGroup(readerGroupName);
+
+        initiateCheckpointExecutor = Executors.newScheduledThreadPool(1);
+        performCheckpointExecutor = Executors.newScheduledThreadPool(1);
+        workerExecutor = Executors.newFixedThreadPool(numWorkers);
+    }
+
+    /**
+     * Commit all transactions that are part of a checkpoint.
+     *
+     * @param checkpointName
+     */
+    private void commitTransactions(String checkpointName) {
```

Comment:
It would be good to have a comment explaining the connection between the checkpoint and the transaction

Reply:
I removed the exactly-once processor from this PR.